README
README for pMARS version 0.9.3 - portable corewar system with ICWS'94 extensions

____________________
What is Core War?

    Core War is a game in which two or more virus-like programs fight against
    each other in a simulated memory space or core.  Core War programs are
    written in an assembly language called Redcode which is interpreted by a
    Core War simulator or MARS (Memory Array Redcode Simulator).  The object of
    the game is to prevent the other program(s) from executing.  For more
    information about Core War check out the usenet newsgroup rec.games.corewar
    and its FAQ list ftp://rtfm.mit.edu/pub/usenet/games/corewar-faq or go to
    http://www.koth.org/.

____________________
pMARS highlights

    * portable, run it on your Mac at home or VAX at work
    * free and comes with source
    * core displays for DOS, Mac and UNIX
    * implements a new redcode dialect, ICWS'94, while remaining compatible
      with ICWS'88
    * powerful redcode extensions: multi-line EQUates, FOR/ROF text repetition
    * one of the fastest simulators written in a high level language
    * full-featured, programmable debugger
    * runs the automated tournament "KotH" at http://www.koth.org and the annual
      ICWS tournaments

____________________
Documentation

    pmars.6 is the nroff-source for UNIX man pages.  You can install it in
    /usr/share/man/man6 for use with the UNIX "man" command or format it with
    nroff -man.  pmars.doc contains man pages without control characters that
    have been formatted for printing.

    doc/primer.94 and doc/primer.cdb contain short introductions to the ICWS'94
    draft and the cdb debugger respectively.  redcode.ref is a quick reference
    to the redcode syntax supported by pMARS.  CONTRIB has guidelines for
    porting pMARS to new platforms and contributing new display code.

____________________
Compiling the source

    There are a number of C preprocessor symbols that control which version
    of pMARS is compiled.  To enable an option, include -DSYMBOLNAME in CFLAGS
    of the makefile or uncomment the relevant section in config.h.

    GRAPHX
        This option enables a platform-specific graphical core display.

    SERVER
        Disables the debugger for a non-interactive tournament version.  The
        pMARS program that runs the KotH email tournaments is compiled with
        SERVER enabled.

    EXT94
        Enables the experimental opcodes SEQ, SNE and NOP, as well as the
        A-field relative addressing modes *, {, and }.  This option should
        usually be enabled.  EXT94 also enables the P-space extensions LDP,
        STP and PIN.

    SMALLMEM
        makes all addresses 16-bit as opposed to the usual 32-bit wide.  This
        limits core size to 65535, but also drastically reduces the memory
        footprint of the program.  We found that SMALLMEM reduces the
        simulation speed of pMARS on most CPUs, with the exception of those
        with a very small primary cache.

    There are other compile directives described in config.h, in particular
    some that fine-tune the UNIX curses display.

    pMARS has been tested with various ANSI and non-ANSI C compilers.  If
    you can't get it to run or you had to change the source extensively,
    contact the authors with a full description of the problems and
    diffs to the source if applicable.

____________________
Platforms

    LINUX
        A standard Makefile is provided.  You can easily specify the desired
	target.  $PREFIX and $MAN_PREFIX control binary and man page install
	locations.  They default to /usr and /usr/share.

	Build and install default graphics display (SVGAlib based).
	# make
	# make install

	You can also customize your build.

	Build curses based display (CURSESGRAPHX).
	# make curses

	Build X-Windows display version (XWINGRAPHX).
	# make xwin

	Build SVGAlib version (GRAPHX).
	# make svga

        The Linux/SVGA version of pMARS has been derived from the DOS
        graphical version and therefore should behave very much like that
        one.

        Some notes for compiling the Linux/SVGA version:
        * You will need the Linux SVGA library (svgalib) version 1.12 or above
            to compile pMARS for Linux (it may work with older libraries, but
            I have not had an opportunity to test it).
        * If you #define GRAPHX in config.h or in the makefile, the graphical
            version will automatically be built unless you explicitely
            specify CURSESGRAPHX.
        * You will have to link with -lvgagl -lvga. makefile already contains
            a sample definition of LIB with these libraries.
        * The code assumes that Function keys etc. map to the 'standard'
            escape sequences. pMARS will not recognize these keys otherwise.
        * You need root privileges for the SVGA library, so either run pMARS
            as root, or a better solution is to set the SUID bit of the
            executable (do the following with root privileges).
            # chown root pmars
            # chmod u+s pmars
        * The code assumes your mouse is available via /dev/mouse.  This is
            usually a link to the 'real' mouse interface, e.g. on my system:
            # cd /dev
            # ls -l mouse
            lrwxrwxrwx   1 root     root        5 Mar  2 00:22 mouse -> ttyS0
        * The second digit of the the argument to -v indicates (just like in
            the DOS version) the graphics mode:
            1 ... 640x480 with 256 colors (SVGA)
            2 ... 800x600 with 256 colors (SVGA)
            3 ...1024x768 with 256 colors (SVGA)
            6 ... 320x200 with 256 colors (VGA)
            all other digits will result in the 320x200x256 mode.

        Deficiencies/bugs of the Linux/SVGA version:
        * Currently, the following keys are recognized by pMARS: F1-F10, the
            cursor keys, insert, delete, home, end, page up and page down,
            and Alt-a to Alt-z.
        * The result of a fight is printed to the console after returning
            from the graphical display -- only the last two lines or so are
            not visible until you press <return>.
        * Bug reports are welcome - just drop a note to akosela@andykosela.com

    DOS
        You will need the free 32-bit DOS compiler DJGPP - a DOS port of gcc -
        and ndmake or some other make program.  pMARS also compiles with the
        16-bit DOS compilers of the Borland/Turbo C family, but the maximum
        number of warriors is limited to 8 because of memory constraints.

        The GRAPHX compile directive enables the combined VGA/textmode
        display of pmarsv.exe.  You can enable the graphics and textmode
        displays selectively with the DOSGRXGRAPHX and DOSTXTGRAPHX
        directives.  If you so desire, you can even link in a curses display
        using the PDCurses library by specifying the CURSESGRAPHX directive.

    MAC
        The Mac GUI version of pMARS, MacpMARS, will compile with Think C
        and MPW C, possibly others.  Source code for the interface/display of
        MacpMARS as well as instructions on how to make the executable are
        in MacpMARS*s.cpt.hqx.  The GUI code has not been updated for v0.7 of
        pMARS yet, so you need the base archive for v0.6 (pmars06s.zip).

    VMS
        pvms*s.zip contains command files to build pMARS for VMS flavors.  You
        need DEC C; VAX C will not work.  This file also contains a complete VMS
        help system.

    OTHERS
        pMARS should compile with Borland C++ for OS/2, although we haven't
        tried it.  The OS/2 version currently doesn't have a core display.
        pMARS has also been reported to compile on Amigas.

____________________
Language support

    All strings are contained in the file str_eng.c for easy translation into
    languages other than english.  If you would like to see pMARS speak your
    native tongue, translate the strings and send the file to us.  We will then
    include the new str_???.c in the next release and might even release the
    foreign language binaries.


ChangeLog
2019-02-08	Andy Kosela	<akosela@andykosela.com>

	- Release 0.9.3
	- Clean up Makefile.
	- Clean up docs.
	- Fix various compiling errors on modern versions of gcc.

2000-12-25	Ilmari Karonen	<iltzu@sci.fi>

	- Release 0.9.2
	- Added -P option to run all possible warrior placements in random
	order (only if compiled with -DPERMUTATE).
	- Added warning if arithmetic overflow occurs during evaluation.
	- Removed upper limit from -F option to prevent KotH abuse.
	- Fixed curdisp.c so that it compiles - patch by Philip Kendall.

2000-10-23	Anton Marsden	<anton@paradise.net.nz>

	- Release 0.9.1
	- Fixed a bug in the multi-warrior code, which was apparently
	introduced in pmars-0.8.5.
	- Added support for a private colourmap if no colours can be
	allocated in X-Windows - patch by Philip Kendall.

2000-08-20	Anton Marsden	<anton@paradise.net.nz>

	- Release 0.9.0
	- New GPL license


src/global.h
/* pMARS -- a portable Memory Array Redcode Simulator
 * Copyright (C) 1993-1996 Albert Ma, Na'ndor Sieben, Stefan Strack and Mintardjo Wangsawidjaja
 * Copyright (C) 2000 Ilmari Karonen
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/*
 * global.h: global header
 * $Id: global.h,v 1.5 2000/12/25 00:56:41 iltzu Exp $
 *
 * 10-23-98 Pentium optimized version 30% faster than the original
 *          Ken Espiritu   (kespirit@cesun1.ce.vt.edu)
 */

#include "config.h"
#include <stdio.h>
#ifndef INT_MAX
#include <limits.h>
#endif
#if defined(__STDC__) || defined(__MSDOS__)
#include <stdlib.h>
#endif

/* *********************************************************************
   System dependent definitions or declarations
   ********************************************************************* */

/* Generic Pointer type */
#if defined(NEW_STYLE)
typedef void *pointer_t;
#else
typedef char *pointer_t;
#endif

#if defined(DOS16)

/* ***** Turbo C FAR model ***** */
#if defined(__TURBOC__)

#include <alloc.h>
#define MALLOC(x)     malloc((size_t)(x))
#define REALLOC(x, y) realloc((pointer_t)(x), (size_t)(y))
#define FREE(x)       free((pointer_t)(x))

#endif                                /* __TURBOC__ */


/* ***** Microsoft/Quick C FAR model ***** */
#if defined(_MSC_VER)

/* Need to test these predefined macros */
#include <malloc.h>

#define MALLOC(x)     _fmalloc((size_t)(x))
#define REALLOC(x, y) _frealloc((pointer_t)(x), (size_t)(y))
#define FREE(x)       _ffree((pointer_t)(x))

#define DOSFARMODEL

#endif                                /* _MSC_VER */

#endif                                /* DOS16 */


/***** Others which don't have MALLOC() defined. *****/

#if !defined(MALLOC)

#define MALLOC(x)     malloc((size_t)(x))
#define REALLOC(x, y) realloc((pointer_t)(x), (size_t)(y))
#define FREE(x)       free((pointer_t)(x))

#endif                                /* ! MALLOC */

/***** Miscellaneous *****/

/* Some old compilers do not have NULL defined */
#ifndef NULL

#ifdef DOSFARMODEL
#define NULL 0L
#else
#define NULL 0
#endif                                /* DOSFARMODEL */

#endif                                /* NULL */

/* unsigned types (renamed to avoid conflict with possibly predefined types) */
typedef unsigned char uChar;
typedef unsigned short uShrt;
typedef unsigned long uLong;

/* FALSE, TRUE */
#ifndef TRUE
enum {
  FALSE, TRUE
};
#endif

/* global error output macro */
#ifdef MACGRAPHX
#define errout(s) macputs(s)
#else
#define errout(s) fputs(s,stderr)
#endif

/* ************************************************************************
   pmars global structures and definitions
   ************************************************************************ */

#if defined(__STDC__)  || defined(DOS16)
#define NEW_STYLE
#endif

/* Version and date */

#define PMARSVER  93
#define PMARSDATE "8/2/2019"

#ifdef VMS                        /* Must change codes to work with VMS error
                                 * handling */
extern  PMARS_FATAL, PMARS_BADCOMLIN, PMARS_PARSEERR;
#define GRAPHERR   1                /* no grphx yet */
#define NOT386     1                /* if this error occurs, I don't wanna hear
                                 * about it. */
#define MEMERR   292                /* %SYS-E-INSFMEM */
#define SERIOUS  &PMARS_FATAL        /* %PMARS-F-FATAL */
#define FNOFOUND 98962                /* %RMS-E-FNF */
#define CLP_NOGOOD &PMARS_BADCOMLIN        /* %PMARS-E-BADCOMLIN */
#define PARSEERR &PMARS_PARSEERR/* ENDABORT */
#define USERABORT 44                /* %SYS-E-ABORT */
#else                                /* everyone else */
/* return code:
   0: success
   Negative number: System error such as insufficient space, etc
   Positive number: User error such as number too big, file not found, etc */
#define GRAPHERR  -4                /* graphic error */
#define NOT386    -3                /* trying to execute 386 code on lesser
                                 * machine */
#define MEMERR    -2                /* insufficient memory, cannot free, etc. */
#define SERIOUS   -1                /* program logic error */
#define FNOFOUND   1                /* File not found */
#define CLP_NOGOOD 2                /* command line argument error */
#define PARSEERR   3                /* File doesn't assemble correctly */
#define USERABORT  4                /* user stopped program from cdb, etc. */
#endif

/* these are used as return codes of internal functions */
#define SUCCESS    0
#define WARNING    0

/* used by eval.c */
#define OVERFLOW  1
#define OK_EXPR   0
#define BAD_EXPR -1
#define DIV_ZERO -2

/* used by cdb.c */
#define NOBREAK 0
#define BREAK   1
#define STEP    2

#define SKIP    1                /* cmdMod settings for cdb: skip next command */
#define RESET   2                /* clear command queue */

#define UNSHARED -1                /* P-space is private */
#define PIN_APPEARED -2

/* used by sim.c and asm.c */
#ifdef NEW_MODES
#define INDIR_A(x) (0x80 & (x))
#define RAW_MODE(x) (0x7F & (x))
#define SYM_TO_INDIR_A(x) ( ((x) - 3) | 0x80)        /* turns index into
                                                 * addr_sym[] to INDIR_A code */
#define INDIR_A_TO_SYM(x) ( RAW_MODE(x) + 3 )        /* vice versa */
#endif

/* used by many */
#define STDOUT stdout

#ifdef DOS16
#define MAXCORESIZE   8192
#else
#ifdef SMALLMEM
#define MAXCORESIZE   65535
#else
#define MAXCORESIZE   ((INT_MAX>>1)+1)
#endif
#endif

#define MAXTASKNUM    INT_MAX
#define MAXROUND      INT_MAX
#define MAXCYCLE      LONG_MAX
#ifdef DOS16
#define MAXWARRIOR         8
#else
#define MAXWARRIOR        36
#endif
#define MAXINSTR         1000

#define MAXSEPARATION MAXCORESIZE/MAXWARRIOR

#define MAXALLCHAR 256

/* The following holds the order in which opcodes, modifiers, and addr_modes
   are represented as in parser. The enumerated field should start from zero */
enum addr_mode {
  IMMEDIATE,                        /* # */
  DIRECT,                        /* $ */
  INDIRECT,                        /* @ */
  PREDECR,                        /* < */
  POSTINC                        /* > */
};

enum op {
  MOV, ADD, SUB, MUL, DIV, MOD, JMZ,
  JMN, DJN, CMP, SLT, SPL, DAT, JMP,
  SEQ,  SNE, NOP, LDP, STP
};                                /* has to match asm.c:opname[] */

enum modifier {
  mA,                                /* .A */
  mB,                                /* .B */
  mAB,                                /* .AB */
  mBA,                                /* .BA */
  mF,                                /* .F */
  mX,                                /* .X */
  mI                                /* .I */
};


#ifdef SMALLMEM
typedef unsigned short ADDR_T;
#define ISNEG(x) ((x)==0xFFFF)        /* use for unsigned ADDR_T */
#else
typedef int ADDR_T;
#define ISNEG(x) ((x)<0)
#endif

typedef unsigned char FIELD_T;
typedef unsigned long U32_T;        /* unsigned long (32 bits) */
typedef long S32_T;


/* Memory structure */
typedef struct mem_struct {
  ADDR_T  A_value, B_value;
  FIELD_T opcode;
  FIELD_T A_mode, B_mode;
  FIELD_T debuginfo;

}       mem_struct;

/* Warrior structure */
typedef struct warrior_struct {
  long    pSpaceIDNumber;
#ifdef DOS16
  ADDR_T far *taskHead, far * taskTail;
#else
  ADDR_T *taskHead, *taskTail;
#endif
  int     tasks;
  ADDR_T  lastResult;
  int     pSpaceIndex;
  ADDR_T  position;                /* load position in core */
  int     instLen;                /* Length of instBank */
  int     offset;                /* Offset value specified by 'ORG' or 'END'.
                                 * 0 is default */
  short   score[MAXWARRIOR * 2 - 1];

  char   *name;                        /* warrior name */
  char   *version;
  char   *date;
  char   *fileName;                /* file name */
  char   *authorName;                /* author name */
  mem_struct *instBank;

  struct warrior_struct *nextWarrior;

}       warrior_struct;

/* ***********************************************************************
   pmars global variable declarations
   *********************************************************************** */

extern int errorcode;
extern int errorlevel;
extern char errmsg[MAXALLCHAR];

/* Some parameters */
extern int warriors;
extern ADDR_T coreSize;
extern int taskNum;
extern ADDR_T instrLim;
extern ADDR_T separation;
extern int rounds;
extern long cycles;

extern int cmdMod;
extern S32_T seed;

extern int SWITCH_b;
extern int SWITCH_e;
extern int SWITCH_k;
extern int SWITCH_8;
extern int SWITCH_f;
extern ADDR_T SWITCH_F;
extern int SWITCH_V;
extern int SWITCH_o;
extern int SWITCH_Q;
extern char *SWITCH_eq;
#ifdef VMS
extern int SWITCH_D;
#endif
#ifdef PERMUTATE
extern int SWITCH_P;
#endif

extern int inCdb;
extern int debugState;
extern int copyDebugInfo;
#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX)
extern int inputRedirection;
#endif
#if defined(XWINGRAPHX)
extern int xWinArgc;
extern char **xWinArgv;
#endif
extern mem_struct INITIALINST;        /* initialize to DAT.F $0,$0 */

extern warrior_struct warrior[MAXWARRIOR];
#ifdef DOS16
extern ADDR_T far *pSpace[MAXWARRIOR];
#else
extern ADDR_T *pSpace[MAXWARRIOR];
#endif
extern ADDR_T pSpaceSize;

/* ***********************************************************************
   display define's, declarations and typedefs
   *********************************************************************** */

#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX)

#if !defined(LINUXGRAPHX)        /* vga.h already defines TEXT to be 0 */
#define         TEXT 0
#endif
#define         GRX 1
#define         SPEEDLEVELS 9
#define         NORMAL_ATTR 0x0700
extern int displayLevel;
extern int displayMode;
extern int displaySpeed;
extern int SWITCH_v;

#if defined(LINUXGRAPHX)        /* needed for correct keyboard handling */
extern struct termios tio_orig;
extern int console_fd;
#endif

#if defined(CURSESGRAPHX)
#include <curses.h>
#if defined(A_NORMAL) || defined(A_BOLD) || defined(A_REVERSE)
#define ATTRIBUTE
#endif                                /* A_* */

extern int refreshInterval;
extern int refIvalAr[SPEEDLEVELS];

#else

extern int keyDelay;
extern int keyDelayAr[SPEEDLEVELS];
extern unsigned long loopDelay;
extern unsigned long loopDelayAr[SPEEDLEVELS];

#endif                                /* CURSESGRAPHX */
#endif                                /* DOSTXTGRAPHX and DOSGRXGRAPHX and
                                 * LINUXGRAPHX */


/* ***********************************************************************
   function prototypes
   *********************************************************************** */

#ifdef NEW_STYLE

extern int
        parse_param(int argc, char *argv[]);
extern int eval_expr(char *expr, long *result);
extern int assemble(char *fName, int aWarrior);
extern void disasm(mem_struct * cells, ADDR_T n, ADDR_T offset);
extern void simulator1(void);
extern char *locview(ADDR_T loc, char *outp);
extern int cdb(char *msg);
extern int score(int warnum);
extern void sort_by_score(int *idxV, int *scrV);
extern int deaths(int warnum);
extern void results(FILE * outp);
extern void sort_by_score();
extern void Exit(int code);
extern void reset_regs(void);
extern void set_reg(char regChr, long val);

#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX)
extern void decode_vopt(int option);
#ifndef LINUXGRAPHX
extern void grputs(char *str);
extern void aputs5(char *str, int attribute);
#endif

#if defined(LINUXGRAPHX)
extern char *svga_gets(char *s, int maxstr);
extern void svga_puts(char *s);
extern void svga_display_close(int wait);
extern void svga_write_menu(void);
extern void svga_open_graphics(void);
extern void svga_clear(void);
extern void svga_update(int curPanel);
extern void svga_clear_arena(void);
extern int svga_getch(void);
#endif

#if defined(XWINGRAPHX)
extern void xWin_puts(char *s);
extern void xWin_write_menu(void);
extern void xWin_clear(void);
extern void xWin_update(int curPanel);
extern void xWin_display_close(int wait);
extern char *xWin_gets(char *s, int maxstr);
extern void xWin_resize(void);
#endif
#endif

#ifdef DOS16
extern char *cellview(mem_struct far * cell, char *outp, int emptyDisp);
#else
extern char *cellview(mem_struct * cell, char *outp, int emptyDisp);
#endif

#else

extern int
        parse_param();
extern int
        eval_expr();
extern int assemble();
extern void disasm();
extern void simulator1();
extern char *locview();
extern char *cellview();
extern int cdb();
extern int score();
extern int deaths();
extern void results();
extern void Exit();
extern void reset_regs();
extern void set_reg();

#if defined(CURSESGRAPHX)
extern void decode_vopt();
extern void aputs5();
#endif

#if defined (LINUXGRAPHX)
extern char *svga_gets();
extern void svga_puts();
extern void svga_display_close();
extern void svga_write_menu();
extern void svga_open_graphics();
extern void svga_clear();
extern void svga_update();
extern void svga_clear_arena();
extern int svga_getch();
#endif

#if defined(XWINGRAPHX)
extern void xWin_puts();
extern void xWin_write_menu();
extern void xWin_clear();
extern void xWin_update();
extern void xWin_display_close();
extern char *xWin_gets();
extern void xWin_resize();
#endif

#endif
